File: backend/services/platform_detector.py
import re
from urllib.parse import urlparse

class PlatformDetector:
    """Detect platform from URL"""
    
    PATTERNS = {
        'youtube': {
            'regex': r'(?:youtube\.com\/watch\?v=|youtu\.be\/|youtube\.com\/shorts\/)([\w-]+)',
            'name': 'YouTube',
            'extractor': 'youtube',
        },
        'instagram': {
            'regex': r'instagram\.com\/(?:p|reel)\/([\w-]+)',
            'name': 'Instagram',
            'extractor': 'instagram',
        },
        'tiktok': {
            'regex': r'(?:tiktok\.com\/.*v=|vm\.tiktok\.com\/)([\w]+)',
            'name': 'TikTok',
            'extractor': 'tiktok',
        },
        'facebook': {
            'regex': r'facebook\.com\/(?:watch\/\?v=|.*\/videos\/)([\d]+)',
            'name': 'Facebook',
            'extractor': 'facebook',
        },
        'reddit': {
            'regex': r'reddit\.com\/r\/\w+\/comments\/(\w+)',
            'name': 'Reddit',
            'extractor': 'reddit',
        },
        'pinterest': {
            'regex': r'pinterest\.com\/pin\/(\d+)',
            'name': 'Pinterest',
            'extractor': 'pinterest',
        },
        'threads': {
            'regex': r'threads\.net\/.*\/(\w+)',
            'name': 'Threads',
            'extractor': 'threads',
        },
    }
    
    @staticmethod
    def detect(url):
        """Detect platform from URL"""
        try:
            urlparse(url)
        except:
            return None
        
        for platform, config in PlatformDetector.PATTERNS.items():
            if re.search(config['regex'], url):
                return {
                    'platform': platform,
                    'name': config['name'],
                    'extractor': config['extractor'],
                }
        
        return None
    
    @staticmethod
    def is_supported(url):
        """Check if URL is supported"""
        return PlatformDetector.detect(url) is not None

File: frontend/src/utils/platform-detector.js
export const platformPatterns = {
  youtube: {
    regex: /(?:youtube\.com\/watch\?v=|youtu\.be\/|youtube\.com\/shorts\/)([\w-]+)/,
    name: 'YouTube',
    icon: 'youtube',
    color: '#FF0000',
  },
  instagram: {
    regex: /instagram\.com\/(?:p|reel)\/([\w-]+)/,
    name: 'Instagram',
    icon: 'instagram',
    color: '#E4405F',
  },
  tiktok: {
    regex: /(?:tiktok\.com\/.*v=|vm\.tiktok\.com\/)(\w+)/,
    name: 'TikTok',
    icon: 'music',
    color: '#000000',
  },
  facebook: {
    regex: /facebook\.com\/(?:watch\/\?v=|.*\/videos\/)(\d+)/,
    name: 'Facebook',
    icon: 'facebook',
    color: '#1877F2',
  },
  reddit: {
    regex: /reddit\.com\/r\/\w+\/comments\/(\w+)/,
    name: 'Reddit',
    icon: 'reddit',
    color: '#FF4500',
  },
  pinterest: {
    regex: /pinterest\.com\/pin\/(\d+)/,
    name: 'Pinterest',
    icon: 'pin',
    color: '#E60023',
  },
  threads: {
    regex: /threads\.net\/.*\/(\w+)/,
    name: 'Threads',
    icon: 'message-circle',
    color: '#000000',
  },
}

export const detectPlatform = (url) => {
  try {
    new URL(url)
  } catch {
    return null
  }

  for (const [key, platform] of Object.entries(platformPatterns)) {
    if (platform.regex.test(url)) {
      return {
        platform: key,
        name: platform.name,
        icon: platform.icon,
        color: platform.color,
      }
    }
  }

  return null
}

export const isValidUrl = (url) => {
  try {
    new URL(url)
    return detectPlatform(url) !== null
  } catch {
    return false
  }
}
